internal/server/routes.go

```diff
@@ -140,43 +147,97 @@ func (s *Server) GenerateRoutes() Routes {
 type HandlerFunc[Req, Res any] func(c *gin.Context, req *Req) (Res, error)
 
 type route[Req, Res any] struct {
-	method              string
-	path                string
-	handler             HandlerFunc[Req, Res]
-	omitFromDocs        bool
-	omitFromTelemetry   bool
-	infraHeaderOptional bool
+	method                     string
+	path                       string
+	handler                    HandlerFunc[Req, Res]
+	omitFromDocs               bool
+	omitFromTelemetry          bool
+	infraVersionHeaderOptional bool
+	noAuthentication           bool
+	noOrgRequired              bool
+}
+
+// TODO: replace this when routes are defined as package-level vars instead of
+// constructed from the get, post, put, del helper functions.
+type routeGroup struct {
+	*gin.RouterGroup
+	noAuthentication bool
+	noOrgRequired    bool
 }
 
-func add[Req, Res any](a *API, r *gin.RouterGroup, route route[Req, Res]) {
-	route.path = path.Join(r.BasePath(), route.path)
+func add[Req, Res any](a *API, group *routeGroup, route route[Req, Res]) {
+	route.path = path.Join(group.BasePath(), route.path)
 
 	if !route.omitFromDocs {
 		a.register(openAPIRouteDefinition(route))
 	}
 
-	wrappedHandler := func(c *gin.Context) {
-		if !route.infraHeaderOptional {
+	route.noAuthentication = group.noAuthentication
+	route.noOrgRequired = group.noOrgRequired
+
+	handler := func(c *gin.Context) {
+		if err := wrapRoute(a, route)(c); err != nil {
+			sendAPIError(c, err)
+		}
+	}
+	bindRoute(a, group.RouterGroup, route.method, route.path, handler)
+}
+
+func wrapRoute[Req, Res any](a *API, route route[Req, Res]) func(*gin.Context) error {
```

Comment:
I would like to see a comment explaining that `wrapRoute` effectively functions as a middleware for requests, there is a lot going on in here

Reply:
Good call, there is a lot going on here. #3104 is a follow up PR that reduces it a bit.

More godoc here would be better. I'll add that now.